csrc/multidevice/pipeline.cpp

```diff
@@ -0,0 +1,319 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ir/builder.h>
+#include <ir/cloner.h>
+#include <ir/printer.h>
+#include <ir/utils.h>
+#include <iter_visitor.h>
+#include <multidevice/pipeline.h>
+#include <multidevice/pipeline_ir.h>
+
+namespace nvfuser {
+
+int PipelineStageDescriptor::running_unique_id_ = 0;
+
+/* Utility class used for Pipeline instantiation called by the Pipeline's
+  constructor. This class is responsible for:
+  - checking that the parameters are valid
+  - analyzing the data dependency from original_fusion_ and descriptor_ to infer
+  what should be the I/O of each stage
+  - creating the IR composing the pipeline:
+       - each StageDescriptor will result in the creation a PipelineStage
+       - each I/O of a PipelineStage will be represented by a PipelineVal
+       - each communication of a PipelineVal between two PipelineStage will
+         be represented by a PipelineCommunication
+  - setting the global I/O of the Pipeline as the PipelineVals representing
+    the global I/O of the original_fusion_
+*/
+class PipelineBuilder final {
+  using ValSet = VectorOfUniqueEntries<Val*>;
+
+ public:
+  PipelineBuilder(Pipeline* pipeline) : pipeline_(pipeline) {
+    validate();
+    fillInfo();
+    buildPipelineVals();
+    buildPipelineStages();
+    buildPipelineCommunications();
+  }
+
+ private:
+  // Points to the pipeline to be constructed
+  Pipeline* pipeline_;
+  // Stores the Stages IR of the pipeline
+  std::vector<PipelineStage*> stages_;
+  // stores the Vals that will be I/O of each stage
+  std::unordered_map<const PipelineStageDescriptor*, ValSet> stage_input_desc_;
+  std::unordered_map<const PipelineStageDescriptor*, ValSet> stage_output_desc_;
+  // Store the stages that consumes a given Val produced at another stage
+  std::unordered_map<Val*, std::vector<const PipelineStageDescriptor*>>
+      val_consumer_stage_desc_;
+  // maps Vals of the original fusion to their corresponding PipelineVals
+  // which are inputs/outputs of a Stage. We need to differentiate those
+  // two cases for when a stage contains only one Val, from which we create
+  // two PipelineVal
+  std::unordered_map<Val*, Val*> val_to_pipeline_val_input_of_stage_;
+  std::unordered_map<Val*, Val*> val_to_pipeline_val_output_of_stage_;
+
+  // Check (partially) that the pipeline is valid and satisfies the assumptions
+  // described in pipeline.h
+  void validate() {
+    std::unordered_set<TensorView*> tv_in_stages;
+    // Check that each Val is a TensorView and that it belongs to at most one
+    // stage
```

Comment:
I had this concern and didn't really know what to do about it, so, for simplicity (when it will come to sharding), I have assumed here that all the Vals are tensorviews -- if not we will get an error.

It shouldn't be too hard to write a patch support to support any Val though. Do you think we should go for it?